clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java

```diff
@@ -195,6 +195,7 @@ public NetworkClientDelegate.PollResult poll(long currentTimeMs) {
             NetworkClientDelegate.UnsentRequest request = makeHeartbeatRequest(currentTimeMs, true);
             // We can ignore the leave response because we can join before or after receiving the response.
             heartbeatRequestState.reset();
+            heartbeatState.reset();
```

Comment:
We might need to reset this at other places where it requires to rejoin.

Reply:
One example would be fenced exception - we need to reset before sending a joining heartbeat.